src/mainboard/clevo/adl-p/ramstage.c

```diff
@@ -3,6 +3,56 @@
 #include <mainboard/gpio.h>
 #include <soc/ramstage.h>
 #include <smbios.h>
+#include <ec/system76/ec/commands.h>
+#include <drivers/efi/efivars.h>
+#include <fmap.h>
+#include <smmstore.h>
+
+#include <Uefi/UefiBaseType.h>
+
+#include <lib.h>
+
+struct fan_point {
+	uint8_t temp;
+	uint8_t duty;
+};
+
+struct fan_curve {
+	struct fan_point point1;
+	struct fan_point point2;
+	struct fan_point point3;
+	struct fan_point point4;
+};
+
+struct smfi_cmd_set_fan_curve {
+	uint8_t fan;
+	struct fan_curve curve;
+};
+
+#define FAN_CURVE_SILENT {      \
+	{ .temp = 0,   .duty = 25  }, \
+	{ .temp = 65,  .duty = 30  }, \
+	{ .temp = 75,  .duty = 35  }, \
+	{ .temp = 100, .duty = 100 }  \
+}
+
+#define FAN_CURVE_PERFORMANCE { \
+	{ .temp = 0,   .duty = 25 },  \
+	{ .temp = 55,  .duty = 35 },  \
+	{ .temp = 75,  .duty = 60 },  \
+	{ .temp = 100, .duty = 100}   \
+}
+
+#define FAN_CURVE_OPTION_SILENT 0
+#define FAN_CURVE_OPTION_PERFORMANCE 1
+
+#define FAN_CURVE_OPTION_DEFAULT FAN_CURVE_OPTION_SILENT
+
+static struct fan_curve fan_curves[] = {
+	[FAN_CURVE_OPTION_SILENT]      = FAN_CURVE_SILENT,
+	[FAN_CURVE_OPTION_PERFORMANCE] = FAN_CURVE_PERFORMANCE
+};
```

Comment:
I think it would be better to have static structs for that instead of arrays and macros:

```
struct fan_curve fan_curve_silent = {
	{ .temp = 0,   .duty = 25  },
	{ .temp = 65,  .duty = 30  },
	{ .temp = 75,  .duty = 35  },
	{ .temp = 100, .duty = 100 } 
}
```

Etc. And then:

```
switch (selection) {
case FAN_CURVE_OPTION_SILENT: cmd.curve = fan_curve_silent; break;
case FAN_CURVE_OPTION_PERFORMANCE: cmd.curve = fan_curve_performance; break;
default: cmd.curve = fan_curve_silent; break;
}
```

Reply:
Also the structs should have `__packed` attribute right? Because the yare used as raw input to EC CMD

Reply:
Okay, done as per suggestion. And yes the structs clearly needed to be packed